pkg/monitortests/kubeapiserver/crdversionchecker/monitortest.go

```diff
@@ -0,0 +1,486 @@
+		CollectedAt: time.Now(),
+		CRDs:        make(map[string]CRDInfo),
+	}
+
+	for _, crd := range crdList.Items {
+		info := crdInfoFromCRD(&crd)
+		snapshot.CRDs[crd.Name] = info
+	}
+
+	return snapshot, nil
+}
+
+// crdInfoFromCRD extracts the relevant information from a CRD object.
+func crdInfoFromCRD(crd *apiextensionsv1.CustomResourceDefinition) CRDInfo {
+	versions := make([]CRDVersionInfo, 0, len(crd.Spec.Versions))
+	for _, v := range crd.Spec.Versions {
+		versions = append(versions, CRDVersionInfo{
+			Name:       v.Name,
+			Served:     v.Served,
+			Storage:    v.Storage,
+			Deprecated: v.Deprecated,
+		})
+	}
+
+	conditions := make([]CRDCondition, 0, len(crd.Status.Conditions))
+	for _, c := range crd.Status.Conditions {
+		conditions = append(conditions, CRDCondition{
+			Type:               string(c.Type),
+			Status:             string(c.Status),
+			Reason:             c.Reason,
+			Message:            c.Message,
+			LastTransitionTime: c.LastTransitionTime.Format(time.RFC3339),
+		})
+	}
+
+	return CRDInfo{
+		Name:           crd.Name,
+		Group:          crd.Spec.Group,
+		Kind:           crd.Spec.Names.Kind,
+		Versions:       versions,
+		StoredVersions: crd.Status.StoredVersions,
+		Conditions:     conditions,
+	}
+}
+
+// buildSummary computes the differences between before and after snapshots.
+func (c *crdVersionChecker) buildSummary() *CRDSummary {
+	summary := &CRDSummary{
+		BeforeUpgrade: *c.beforeSnapshot,
+		AfterUpgrade:  *c.afterSnapshot,
+		AddedCRDs:     []string{},
+		RemovedCRDs:   []string{},
+		ChangedCRDs:   []CRDChangeSummary{},
+	}
+
+	// Find added CRDs (in after but not in before)
+	for name := range c.afterSnapshot.CRDs {
+		if _, exists := c.beforeSnapshot.CRDs[name]; !exists {
+			summary.AddedCRDs = append(summary.AddedCRDs, name)
+		}
+	}
+	sort.Strings(summary.AddedCRDs)
+
+	// Find removed CRDs (in before but not in after)
+	for name := range c.beforeSnapshot.CRDs {
+		if _, exists := c.afterSnapshot.CRDs[name]; !exists {
+			summary.RemovedCRDs = append(summary.RemovedCRDs, name)
+		}
+	}
+	sort.Strings(summary.RemovedCRDs)
+
+	// Find changed CRDs (exist in both but have version differences)
+	for name, beforeCRD := range c.beforeSnapshot.CRDs {
+		afterCRD, exists := c.afterSnapshot.CRDs[name]
+		if !exists {
+			continue // Already handled as removed
+		}
+
+		changeSummary := computeCRDChanges(name, beforeCRD, afterCRD)
+		if changeSummary != nil {
+			summary.ChangedCRDs = append(summary.ChangedCRDs, *changeSummary)
+		}
+	}
+
+	// Sort changed CRDs by name for consistent output
+	sort.Slice(summary.ChangedCRDs, func(i, j int) bool {
+		return summary.ChangedCRDs[i].Name < summary.ChangedCRDs[j].Name
+	})
+
+	return summary
+}
+
+// computeCRDChanges determines what changed between two versions of a CRD.
+func computeCRDChanges(name string, before, after CRDInfo) *CRDChangeSummary {
+	beforeVersions := make(map[string]bool)
+	afterVersions := make(map[string]bool)
```

Comment:
Nit, if you're using this as just a set, `map[string]struct{}` is more idiomatic and more efficient